src/SimplePie.php

```diff
@@ -1900,7 +1900,7 @@ protected function fetch_data(&$cache)
                     $this->data = [];
                 }
                 // Check if the cache has been updated
-                elseif (isset($this->data['cache_expiration_time']) && $this->data['cache_expiration_time'] > time()) {
+                elseif (!isset($this->data['cache_expiration_time']) || $this->data['cache_expiration_time'] < time()) {
```

Comment:
Reading the explanations from both of you I think this PR is reasonable.

The last remaining thing not clear to me is how it can this patch fix anything without https://github.com/FreshRSS/simplepie/blob/5e92ce767597aa9f5cbb00d88c7591b2e604dc5f/src/Cache/BaseDataCache.php#L58-L62

If we use the same `$this->cache_duration` as TTL passed to `$cache->set_data()` and to `$this->data['cache_expiration_time']`, I would expect that if the cache expires, the following would return the default empty array, and the above line would never be executed, as it is within `if (!empty($this->data))`:

https://github.com/simplepie/simplepie/blob/6b97dec75ed45c11a7411616643f855fe13c2add/src/SimplePie.php#L1874

Reply:
> how it can this patch fix anything

This patch will fix some cases when the internal TTL has not expired yet. I think this is needed in any case.

But I completely agree that to have a working 304 cache, more fixes of the logic are needed. This PR focusses on the straightforward bugs in the current logic, which are needed no matter what.

Reply:
But how could the internal TTL not have expired when `$this->data['cache_expiration_time']` did – are not they set to the same cache duration?

Reply:
While the default cache implementation ("legacy file cache") is still broken, I believe passing another implementation might work (but I have not investigated in details)

Reply:
Ah, alright. We can set the TTL to `null` in all `set_data` calls in a follow up PR to actually fix it then.